File: frontend/src/engine/flotation.ts
/**
 * Flotation circuit: rougher, optional regrind, cleaner and optional recleaner, with recycles (port of
 * engine/flotation.py; theory in docs/methodologies/05_flotation.md). Per perfectly mixed cell a class is
 * recovered with r = (k tau + ENT w)/(1 + k tau + ENT w); k = 60 P Sb f_size f_dose with Sb = 6 Jg/D32;
 * ENT after Savassi et al. (1998). The recycle is a fixed point on an absolute and a per-class relative change.
 */
import { constant } from './constants';
import { grid, type Vec } from './grid';
import type { Bank, Flags, FlotationPlant, OperatingPoint } from './model';
import type { ResolvedOre } from './ore';
import { speciesContentOf, speciesDefs, toMinerals, toSpecies, type Species, type SpeciesDef } from './species';
import { Stream } from './streams';

export function bubbleSurfaceFlux(jgCmS: number, plant: FlotationPlant): number {
  const d32 = plant.d32_base_mm + plant.d32_slope_mm_per_cm_s * jgCmS;
  return constant('flotation.sb_factor') * jgCmS / d32;
}

/** ENT = 2/(exp(z) + exp(-z)), z = 2.292 (d/xi)^adj, adj = 1 - ln(1/delta)/exp(d/xi), exponents capped. */
export function savassiEntrainment(xiUm: number, delta: number): Vec {
  const size = grid().size;
  const cap = Math.log(Number.MAX_VALUE) / 2.0;
  const out = new Float64Array(size.length);
  for (let i = 0; i < size.length; i += 1) {
    const ratio = size[i] / xiUm;
    const adj = 1.0 - Math.log(1.0 / delta) / Math.exp(Math.min(ratio, cap));
    const z = Math.min(constant('savassi.constant') * Math.pow(ratio, adj), cap);
    out[i] = 2.0 / (Math.exp(z) + Math.exp(-z));
  }
  return out;
}

export function sizeResponse(optimumUm: number, fineWidth: number, coarseWidth: number): Vec {
  const size = grid().size;
  const out = new Float64Array(size.length);
  for (let i = 0; i < size.length; i += 1) {
    const logRatio = Math.log(size[i] / optimumUm);
    const width = logRatio < 0.0 ? fineWidth : coarseWidth;
    out[i] = Math.exp(-0.5 * Math.pow(logRatio / width, 2));
  }
  return out;
}

export function doseResponse(doseGpt: number, halfDoseGpt: number, unresponsive: number): number {
  return unresponsive + (1.0 - unresponsive) * doseGpt / (doseGpt + halfDoseGpt);
}

/** True-flotation rate constant (1/min) of every particle class by size. */
export function rateConstants(ore: ResolvedOre, defs: SpeciesDef[], sb: number, doseGpt: number): Species {
  const perMinute = constant('time.seconds_per_minute');
  const n = grid().n;
  const out: Species = {};
  for (const d of defs) {
    const response = ore.spec[d.mineral].flotation;
    if (response === null) { out[d.id] = new Float64Array(n); continue; }
    let p = response.floatability;
    if (d.kind === 'composite') p = p * Math.pow(ore.spec[d.mineral].composite_content, constant('flotation.composite_surface_exponent'));
    const shape = sizeResponse(response.optimum_size_um, response.fine_width, response.coarse_width);
    const dose = doseResponse(doseGpt, response.half_dose_gpt, response.unresponsive_fraction);
    const k = new Float64Array(n);
    for (let i = 0; i < n; i += 1) k[i] = perMinute * p * sb * shape[i] * dose;
    out[d.id] = k;
  }
  return out;
}

export type BankResult = {
  recovery: Species;
  water_recovery: number;
  cell_water_recovery: number;
  tau_cell_min: number;
  residence_min: number;
  entrained_share: Species;
  cell_recovery: Species;
};

const vsum = (v: Vec) => { let s = 0.0; for (let i = 0; i < v.length; i += 1) s += v[i]; return s; };

export function pulpFlowM3Min(species: Species, defs: SpeciesDef[], waterTph: number): number {
  const waterDensity = constant('water.density_t_m3');
  const minutes = constant('time.minutes_per_hour');
  let solids = 0.0;
  for (const d of defs) solids += vsum(species[d.id]) / d.density;
  return (solids + waterTph / waterDensity) / minutes;
}

export function bank(rates: Species, ent: Vec, cells: number, bankDef: Bank, flowM3Min: number, sb: number, waterFloatability: number): BankResult {
  const perMinute = constant('time.seconds_per_minute');
  const tau = bankDef.cell_volume_m3 * (1.0 - bankDef.gas_holdup) / flowM3Min;
  const kw = perMinute * waterFloatability * sb;
  const rw = kw * tau / (1.0 + kw * tau);
  const w = rw / (1.0 - rw);
  const recovery: Species = {};
  const share: Species = {};
  const cell: Species = {};
  for (const [key, k] of Object.entries(rates)) {
    const rec = new Float64Array(k.length);
    const sh = new Float64Array(k.length);
    const ce = new Float64Array(k.length);
    for (let i = 0; i < k.length; i += 1) {
      const entEff = ent[i] * bankDef.wash_factor;
      const numerator = k[i] * tau + entEff * w;
      const r = numerator / (1.0 + numerator);
      ce[i] = r;
      rec[i] = 1.0 - Math.pow(1.0 - r, cells);
      sh[i] = numerator > 0.0 ? entEff * w / numerator : 0.0;
    }
    recovery[key] = rec; share[key] = sh; cell[key] = ce;
  }
  return { recovery, water_recovery: 1.0 - Math.pow(1.0 - rw, cells), cell_water_recovery: rw, tau_cell_min: tau,
    residence_min: tau * cells, entrained_share: share, cell_recovery: cell };
}

export type FlotationResult = {
  streams: Record<string, Stream>;
  species_final: Species;
  rougher: BankResult;
  cleaner: BankResult;
  recleaner: BankResult | null;
  rougher_feed_species: Species;
  feed_species: Species;
  defs: SpeciesDef[];
  iterations: number;
  residual_tph: number;
  sb_rougher: number;
  sb_cleaner: number;
  rates_rougher: Species;
  dilution_water_tph: number;
  regrind_feed_tph: number;
  dilution_cleaner_tph: number;
  dilution_recleaner_tph: number;
  dilution_rougher_tph: number;
  relative_change: number;
};

function solidsOf(species: Species): number {
  let s = 0.0;
  for (const v of Object.values(species)) s += vsum(v);
  return s;
}

/** Water after diluting to the target solids fraction; water is only ever added. */
function diluted(water: number, solids: number, targetSolids: number): number {
  return Math.max(water, solids * (1.0 - targetSolids) / targetSolids);
}

/** Largest change of any particle class between passes, relative to that class's total flow. */
function relativeChange(next: Species, prev: Species): number {
  let worst = 0.0;
  for (const [k, values] of Object.entries(next)) {
    let scale = 0.0;
    let diff = 0.0;
    for (let i = 0; i < values.length; i += 1) { scale += Math.abs(values[i]); diff += Math.abs(values[i] - prev[k][i]); }
    if (scale > 0.0) worst = Math.max(worst, diff / scale);
  }
  return worst;
}

const mapSpecies = (keys: string[], fn: (k: string) => Vec): Species => {
  const out: Species = {};
  for (const k of keys) out[k] = fn(k);
  return out;
};
const times = (a: Vec, b: Vec) => { const o = new Float64Array(a.length); for (let i = 0; i < a.length; i += 1) o[i] = a[i] * b[i]; return o; };
const plus = (a: Vec, b: Vec) => { const o = new Float64Array(a.length); for (let i = 0; i < a.length; i += 1) o[i] = a[i] + b[i]; return o; };
const minus = (a: Vec, b: Vec) => { const o = new Float64Array(a.length); for (let i = 0; i < a.length; i += 1) o[i] = a[i] - b[i]; return o; };

type PassState = {
  rougher: BankResult; conc_r: Species; water_conc_r: number; ground: Species; y: Species; water_y: number; dilution_c: number;
  cleaner: BankResult; conc_c: Species; tail_c: Species; water_conc_c: number; water_tail_c: number;
  recleaner: BankResult | null; water_z: number; dilution_rc: number; final: Species; water_final: number; tail_rc: Species; water_tail_rc: number;
};

export function runFlotation(fo: Species, feedWaterIn: number, ore: ResolvedOre, plant: FlotationPlant, op: OperatingPoint, flags: Flags,
  regrind: ((minerals: Record<string, Vec>) => Record<string, Vec>) | null = null): FlotationResult {
  const defs = speciesDefs(ore);
  const keys = Object.keys(fo);
  const arrivingWater = feedWaterIn;
  let feedWater = feedWaterIn;
  if (plant.rougher.feed_solids > 0.0) feedWater = diluted(feedWater, solidsOf(fo), plant.rougher.feed_solids);
  const rougherDilution = feedWater - arrivingWater;
  const ent = savassiEntrainment(plant.entrainment_size_um, plant.drainage);
  const sbR = bubbleSurfaceFlux(op.jg_cm_s, plant);
  const sbC = bubbleSurfaceFlux(plant.cleaner.jg_cm_s, plant);
  const kR = rateConstants(ore, defs, sbR, op.collector_gpt);
  const kC = rateConstants(ore, defs, sbC, op.collector_gpt);
  const recl = plant.recleaner;
  const sbRc = recl !== null ? bubbleSurfaceFlux(recl.jg_cm_s, plant) : 0.0;
  const kRc = recl !== null ? rateConstants(ore, defs, sbRc, op.collector_gpt) : {};
  const n = grid().n;
  const zeros = mapSpecies(defs.map(d => d.id), () => new Float64Array(n));

  const passOnce = (x: Species, waterX: number, tRc: Species, waterTRc: number): PassState => {
    const rougher = bank(kR, ent, op.rougher_cells, plant.rougher, pulpFlowM3Min(x, defs, waterX), sbR, plant.water_floatability);
    const concR = mapSpecies(keys, k => times(rougher.recovery[k], x[k]));
    const waterConcR = rougher.water_recovery * waterX;
    const ground = regrind !== null ? toSpecies(new Stream(regrind(toMinerals(concR, defs, ore)), waterConcR), ore, defs) : concR;
    const y = mapSpecies(keys, k => plus(ground[k], tRc[k]));
    const waterY0 = waterConcR + waterTRc;
    const waterY = diluted(waterY0, solidsOf(y), plant.cleaner.feed_solids);
    const cleaner = bank(kC, ent, plant.cleaner.cells, plant.cleaner, pulpFlowM3Min(y, defs, waterY), sbC, plant.water_floatability);
    const concC = mapSpecies(keys, k => times(cleaner.recovery[k], y[k]));
    const tailC = mapSpecies(keys, k => minus(y[k], concC[k]));
    const waterConcC = cleaner.water_recovery * waterY;
    const base = { rougher, conc_r: concR, water_conc_r: waterConcR, ground, y, water_y: waterY, dilution_c: waterY - waterY0, cleaner,
      conc_c: concC, tail_c: tailC, water_conc_c: waterConcC, water_tail_c: waterY - waterConcC };
    if (recl !== null) {
      const waterZ = diluted(waterConcC, solidsOf(concC), recl.feed_solids);
      const recleaner = bank(kRc, ent, recl.cells, recl, pulpFlowM3Min(concC, defs, waterZ), sbRc, plant.water_floatability);
      const final = mapSpecies(keys, k => times(recleaner.recovery[k], concC[k]));
      const waterFinal = recleaner.water_recovery * waterZ;
      return { ...base, recleaner, water_z: waterZ, dilution_rc: waterZ - waterConcC, final, water_final: waterFinal,
        tail_rc: mapSpecies(keys, k => minus(concC[k], final[k])), water_tail_rc: waterZ - waterFinal };
    }
    return { ...base, recleaner: null, water_z: 0.0, final: concC, water_final: waterConcC, tail_rc: zeros, water_tail_rc: 0.0, dilution_rc: 0.0 };
  };

  let x = mapSpecies(keys, k => Float64Array.from(fo[k]));
  let waterX = feedWater;
  let tRc = zeros;
  let waterTRc = 0.0;
  const tol = constant('numerics.recycle_tolerance_tph');
  const rtol = constant('numerics.recycle_rel_tolerance');
  let iterations = 0;
  let residual = Infinity;
  let relative = Infinity;
  const maxIterations = constant('numerics.recycle_max_iterations');
  for (iterations = 1; iterations <= maxIterations; iterations += 1) {
    const state = passOnce(x, waterX, tRc, waterTRc);
    const xNew = mapSpecies(keys, k => plus(fo[k], state.tail_c[k]));
    const waterXNew = feedWater + state.water_tail_c;
    const tRcNew = state.tail_rc;
    const waterTRcNew = state.water_tail_rc;
    let maxX = 0.0;
    let maxT = 0.0;
    for (const k of keys) {
      for (let i = 0; i < n; i += 1) {
        maxX = Math.max(maxX, Math.abs(xNew[k][i] - x[k][i]));
        maxT = Math.max(maxT, Math.abs(tRcNew[k][i] - tRc[k][i]));
      }
    }
    residual = maxX + Math.abs(waterXNew - waterX) + maxT + Math.abs(waterTRcNew - waterTRc);
    relative = Math.max(relativeChange(xNew, x), relativeChange(tRcNew, tRc));
    x = xNew; waterX = waterXNew; tRc = tRcNew; waterTRc = waterTRcNew;
    if (residual < tol && relative < rtol) break;
  }
  if (iterations > maxIterations) iterations = maxIterations;
  if (residual >= tol || relative >= rtol) {
    flags.add('recycle_not_converged', `Flotation recycle residual ${residual.toExponential(2)} t/h (relative ${relative.toExponential(2)}) after ${iterations} iterations.`);
  }
  const s = passOnce(x, waterX, tRc, waterTRc);
  const stream = (species: Species, water: number) => new Stream(toMinerals(species, defs, ore), water);
  const streams: Record<string, Stream> = {
    flotation_feed: stream(fo, feedWater),
    rougher_feed: stream(x, waterX),
    rougher_concentrate: stream(s.conc_r, s.water_conc_r),
    rougher_tail: stream(mapSpecies(keys, k => minus(x[k], s.conc_r[k])), waterX - s.water_conc_r),
  };
  if (regrind !== null) streams.regrind_product = stream(s.ground, s.water_conc_r);
  streams.cleaner_feed = stream(s.y, s.water_y);
  streams.cleaner_concentrate = stream(s.conc_c, s.water_conc_c);
  streams.cleaner_tail = stream(s.tail_c, s.water_tail_c);
  if (recl !== null) {
    streams.recleaner_feed = stream(s.conc_c, s.water_z);
    streams.recleaner_concentrate = stream(s.final, s.water_final);
    streams.recleaner_tail = stream(s.tail_rc, s.water_tail_rc);
  }
  return {
    streams, species_final: s.final, rougher: s.rougher, cleaner: s.cleaner, recleaner: s.recleaner, rougher_feed_species: x,
    feed_species: fo, defs, iterations, residual_tph: residual, sb_rougher: sbR, sb_cleaner: sbC, rates_rougher: kR,
    dilution_water_tph: rougherDilution + s.dilution_c + s.dilution_rc, regrind_feed_tph: solidsOf(s.conc_r),
    dilution_cleaner_tph: s.dilution_c, dilution_recleaner_tph: s.dilution_rc, dilution_rougher_tph: rougherDilution,
    relative_change: relative,
  };
}

export function finalStreamName(result: FlotationResult): string {
  return result.recleaner !== null ? 'recleaner_concentrate' : 'cleaner_concentrate';
}

/** Cumulative grade and recovery of `species` along the rougher cells (grade-recovery curve). */
export function bankProfile(result: FlotationResult, ore: ResolvedOre, species: string, cellsInBank: number): Array<Record<string, number>> {
  const x = result.rougher_feed_species;
  const defs = result.defs;
  const cells = result.rougher.cell_recovery;
  const content: Record<string, number> = {};
  for (const d of defs) content[d.id] = speciesContentOf(d, ore, species);
  let total = 0.0;
  let feedMass = 0.0;
  for (const d of defs) { total += content[d.id] * vsum(x[d.id]); feedMass += vsum(x[d.id]); }
  const out: Array<Record<string, number>> = [];
  for (let j = 1; j <= cellsInBank; j += 1) {
    let mass = 0.0;
    let value = 0.0;
    for (const d of defs) {
      let recovered = 0.0;
      for (let i = 0; i < x[d.id].length; i += 1) recovered += x[d.id][i] * (1.0 - Math.pow(1.0 - cells[d.id][i], j));
      mass += recovered;
      value += content[d.id] * recovered;
    }
    out.push({ cell: j, recovery: total > 0.0 ? value / total : 0.0, grade: mass > 0.0 ? value / mass : 0.0,
      mass_pull: feedMass > 0.0 ? mass / feedMass : 0.0 });
  }
  return out;
}

File: data-pipeline/pipeline/engine/flotation.py
"""Flotation circuit: rougher, optional regrind, cleaner and optional recleaner, with recycles.

Per perfectly mixed cell, with residence ``tau`` (min), true-flotation rate ``k`` (1/min), degree of
entrainment ``ENT`` and water recovery ``r_w`` (``w = r_w/(1 - r_w)``), a particle class is
recovered with ``r = (k tau + ENT w)/(1 + k tau + ENT w)``; a bank of N equal cells recovers
``1 - (1 - r)^N``, which reduces to ``1 - (N/(N + k tau_bank))^N`` without entrainment.
``k = 60 P Sb f_size f_dose`` with ``Sb = 6 Jg/D32`` (Gorain et al. 1997); ENT after Savassi et al.
(1998). Cleaner tails return to the rougher feed and recleaner tails to the cleaner feed; the
regrind (open-circuit population balance) changes sizes, and so liberation, before cleaning.
Sources and assumptions: docs/methodologies/05_flotation.md.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Callable

import numpy as np

from .constants import constant
from .grid import grid
from .model import Bank, Flags, FlotationPlant, OperatingPoint
from .ore import ResolvedOre
from .species import SpeciesDef, species_content, species_defs, to_minerals, to_species
from .streams import Stream

Species = dict[str, np.ndarray]


def bubble_surface_flux(jg_cm_s: float, plant: FlotationPlant) -> float:
    d32 = plant.d32_base_mm + plant.d32_slope_mm_per_cm_s * jg_cm_s
    return float(constant("flotation.sb_factor")) * jg_cm_s / d32


def savassi_entrainment(xi_um: float, delta: float) -> np.ndarray:
    """``ENT = 2 / (exp(z) + exp(-z))``, ``z = 2.292 (d/xi)^adj``, ``adj = 1 - ln(1/delta)/exp(d/xi)``.

    Exponents are capped where ENT is already zero in double precision, so coarse classes do not
    overflow.
    """
    size = grid().size
    ratio = size / xi_um
    cap = math.log(np.finfo(float).max) / 2.0
    adj = 1.0 - math.log(1.0 / delta) / np.exp(np.minimum(ratio, cap))
    z = np.minimum(float(constant("savassi.constant")) * np.power(ratio, adj), cap)
    return 2.0 / (np.exp(z) + np.exp(-z))


def size_response(optimum_um: float, fine_width: float, coarse_width: float) -> np.ndarray:
    size = grid().size
    log_ratio = np.log(size / optimum_um)
    width = np.where(log_ratio < 0.0, fine_width, coarse_width)
    return np.exp(-0.5 * np.square(log_ratio / width))


def dose_response(dose_gpt: float, half_dose_gpt: float, unresponsive: float) -> float:
    return unresponsive + (1.0 - unresponsive) * dose_gpt / (dose_gpt + half_dose_gpt)


def rate_constants(ore: ResolvedOre, defs: list[SpeciesDef], sb: float, dose_gpt: float) -> Species:
    """True-flotation rate constant (1/min) of every particle class by size."""
    per_minute = float(constant("time.seconds_per_minute"))
    out: Species = {}
    for d in defs:
        response = ore.spec[d.mineral].flotation
        if response is None:
            out[d.id] = np.zeros(grid().n)
            continue
        p = response.floatability
        if d.kind == "composite":
            p = p * ore.spec[d.mineral].composite_content ** float(constant("flotation.composite_surface_exponent"))
        shape = size_response(response.optimum_size_um, response.fine_width, response.coarse_width)
        out[d.id] = per_minute * p * sb * shape * dose_response(dose_gpt, response.half_dose_gpt, response.unresponsive_fraction)
    return out


@dataclass
class BankResult:
    recovery: Species
    water_recovery: float
    cell_water_recovery: float
    tau_cell_min: float
    residence_min: float
    entrained_share: Species
    cell_recovery: Species


def pulp_flow_m3_min(species: Species, defs: list[SpeciesDef], water_tph: float) -> float:
    water_density = float(constant("water.density_t_m3"))
    minutes = float(constant("time.minutes_per_hour"))
    solids = sum(float(np.sum(species[d.id])) / d.density for d in defs)
    return (solids + water_tph / water_density) / minutes


def bank(rates: Species, ent: np.ndarray, cells: int, bank_def: Bank, flow_m3_min: float,
         sb: float, water_floatability: float) -> BankResult:
    per_minute = float(constant("time.seconds_per_minute"))
    tau = bank_def.cell_volume_m3 * (1.0 - bank_def.gas_holdup) / flow_m3_min
    kw = per_minute * water_floatability * sb
    rw = kw * tau / (1.0 + kw * tau)
    w = rw / (1.0 - rw)
    ent_eff = ent * bank_def.wash_factor
    recovery, share, cell = {}, {}, {}
    for key, k in rates.items():
        numerator = k * tau + ent_eff * w
        r = numerator / (1.0 + numerator)
        cell[key] = r
        recovery[key] = 1.0 - np.power(1.0 - r, cells)
        share[key] = np.where(numerator > 0.0, ent_eff * w / np.where(numerator > 0.0, numerator, 1.0), 0.0)
    return BankResult(recovery, 1.0 - (1.0 - rw) ** cells, rw, tau, tau * cells, share, cell)


@dataclass
class FlotationResult:
    streams: dict[str, Stream]
    species_final: Species
    rougher: BankResult
    cleaner: BankResult
    recleaner: BankResult | None
    rougher_feed_species: Species
    feed_species: Species
    defs: list[SpeciesDef]
    iterations: int
    residual_tph: float
    sb_rougher: float
    sb_cleaner: float
    rates_rougher: Species
    dilution_water_tph: float
    regrind_feed_tph: float
    dilution_cleaner_tph: float
    dilution_recleaner_tph: float
    dilution_rougher_tph: float


def _solids(species: Species) -> float:
    return float(sum(float(np.sum(v)) for v in species.values()))


def _diluted(water: float, solids: float, target_solids: float) -> float:
    """Water after diluting to the target solids fraction; water is only ever added."""
    return max(water, solids * (1.0 - target_solids) / target_solids)


def run_flotation(fo: Species, feed_water: float, ore: ResolvedOre, plant: FlotationPlant, op: OperatingPoint, flags: Flags,
                  regrind: Callable[[dict[str, np.ndarray]], dict[str, np.ndarray]] | None = None) -> FlotationResult:
    """Flotation of particle classes ``fo`` (t/h by class and size) carried forward from the upstream
    separation; only the regrind product is split again into classes, because it is a product of
    breakage."""
    defs = species_defs(ore)
    arriving_water = feed_water
    if plant.rougher.feed_solids > 0.0:
        feed_water = _diluted(feed_water, _solids(fo), plant.rougher.feed_solids)
    rougher_dilution = feed_water - arriving_water
    ent = savassi_entrainment(plant.entrainment_size_um, plant.drainage)
    sb_r = bubble_surface_flux(op.jg_cm_s, plant)
    sb_c = bubble_surface_flux(plant.cleaner.jg_cm_s, plant)
    k_r = rate_constants(ore, defs, sb_r, op.collector_gpt)
    k_c = rate_constants(ore, defs, sb_c, op.collector_gpt)
    recl = plant.recleaner
    sb_rc = bubble_surface_flux(recl.jg_cm_s, plant) if recl is not None else 0.0
    k_rc = rate_constants(ore, defs, sb_rc, op.collector_gpt) if recl is not None else {}
    zeros = {d.id: np.zeros(grid().n) for d in defs}

    def pass_once(x: Species, water_x: float, t_rc: Species, water_t_rc: float) -> dict[str, object]:
        rougher = bank(k_r, ent, op.rougher_cells, plant.rougher, pulp_flow_m3_min(x, defs, water_x), sb_r, plant.water_floatability)
        conc_r = {k: rougher.recovery[k] * x[k] for k in x}
        water_conc_r = rougher.water_recovery * water_x
        if regrind is not None:
            ground = to_species(Stream(regrind(to_minerals(conc_r, defs, ore)), water_conc_r), ore, defs, flags)
        else:
            ground = conc_r
        y = {k: ground[k] + t_rc[k] for k in x}
        water_y0 = water_conc_r + water_t_rc
        water_y = _diluted(water_y0, _solids(y), plant.cleaner.feed_solids)
        cleaner = bank(k_c, ent, plant.cleaner.cells, plant.cleaner, pulp_flow_m3_min(y, defs, water_y), sb_c, plant.water_floatability)
        conc_c = {k: cleaner.recovery[k] * y[k] for k in x}
        tail_c = {k: y[k] - conc_c[k] for k in x}
        water_conc_c = cleaner.water_recovery * water_y
        out: dict[str, object] = {"rougher": rougher, "conc_r": conc_r, "water_conc_r": water_conc_r, "ground": ground,
                                  "y": y, "water_y": water_y, "dilution_c": water_y - water_y0, "cleaner": cleaner,
                                  "conc_c": conc_c, "tail_c": tail_c, "water_conc_c": water_conc_c,
                                  "water_tail_c": water_y - water_conc_c}
        if recl is not None:
            water_z = _diluted(water_conc_c, _solids(conc_c), recl.feed_solids)
            recleaner = bank(k_rc, ent, recl.cells, recl, pulp_flow_m3_min(conc_c, defs, water_z), sb_rc, plant.water_floatability)
            final = {k: recleaner.recovery[k] * conc_c[k] for k in x}
            water_final = recleaner.water_recovery * water_z
            out.update({"recleaner": recleaner, "water_z": water_z, "dilution_rc": water_z - water_conc_c, "final": final,
                        "water_final": water_final, "tail_rc": {k: conc_c[k] - final[k] for k in x},
                        "water_tail_rc": water_z - water_final})
        else:
            out.update({"recleaner": None, "final": conc_c, "water_final": water_conc_c, "tail_rc": zeros,
                        "water_tail_rc": 0.0, "dilution_rc": 0.0})
        return out

    x = {k: v.copy() for k, v in fo.items()}
    water_x = feed_water
    t_rc, water_t_rc = zeros, 0.0
    tol = float(constant("numerics.recycle_tolerance_tph"))
    iterations, residual = 0, math.inf
    for iterations in range(1, int(constant("numerics.recycle_max_iterations")) + 1):
        state = pass_once(x, water_x, t_rc, water_t_rc)
        x_new = {k: fo[k] + state["tail_c"][k] for k in x}
        water_x_new = feed_water + state["water_tail_c"]
        t_rc_new = state["tail_rc"]
        water_t_rc_new = state["water_tail_rc"]
        residual = (max(float(np.max(np.abs(x_new[k] - x[k]))) for k in x) + abs(water_x_new - water_x)
                    + max(float(np.max(np.abs(t_rc_new[k] - t_rc[k]))) for k in x) + abs(water_t_rc_new - water_t_rc))
        x, water_x, t_rc, water_t_rc = x_new, water_x_new, t_rc_new, water_t_rc_new
        if residual < tol:
            break
    if residual >= tol:
        flags.add("recycle_not_converged", f"Flotation recycle residual {residual:.2e} t/h after {iterations} iterations.")
    s = pass_once(x, water_x, t_rc, water_t_rc)

    def stream(species: Species, water: float) -> Stream:
        return Stream(to_minerals(species, defs, ore), water)

    rougher: BankResult = s["rougher"]
    streams = {
        "flotation_feed": stream(fo, feed_water),
        "rougher_feed": stream(x, water_x),
        "rougher_concentrate": stream(s["conc_r"], s["water_conc_r"]),
        "rougher_tail": stream({k: x[k] - s["conc_r"][k] for k in x}, water_x - s["water_conc_r"]),
    }
    if regrind is not None:
        streams["regrind_product"] = stream(s["ground"], s["water_conc_r"])
    streams.update({
        "cleaner_feed": stream(s["y"], s["water_y"]),
        "cleaner_concentrate": stream(s["conc_c"], s["water_conc_c"]),
        "cleaner_tail": stream(s["tail_c"], s["water_tail_c"]),
    })
    if recl is not None:
        streams.update({
            "recleaner_feed": stream(s["conc_c"], s["water_z"]),
            "recleaner_concentrate": stream(s["final"], s["water_final"]),
            "recleaner_tail": stream(s["tail_rc"], s["water_tail_rc"]),
        })
    return FlotationResult(streams, s["final"], rougher, s["cleaner"], s["recleaner"], x, fo, defs, iterations, residual,
                           sb_r, sb_c, k_r, rougher_dilution + s["dilution_c"] + s["dilution_rc"], _solids(s["conc_r"]),
                           s["dilution_c"], s["dilution_rc"], rougher_dilution)


def final_stream_name(result: FlotationResult) -> str:
    return "recleaner_concentrate" if result.recleaner is not None else "cleaner_concentrate"


def bank_profile(result: FlotationResult, ore: ResolvedOre, species: str, cells_in_bank: int) -> list[dict[str, float]]:
    """Cumulative grade and recovery of ``species`` along the rougher cells (grade-recovery curve)."""
    x, defs, cells = result.rougher_feed_species, result.defs, result.rougher.cell_recovery
    content = {d.id: species_content(d, ore, species) for d in defs}
    total = sum(content[d.id] * float(np.sum(x[d.id])) for d in defs)
    feed_mass = sum(float(np.sum(x[d.id])) for d in defs)
    out = []
    for j in range(1, cells_in_bank + 1):
        mass = sum(float(np.sum(x[d.id] * (1.0 - np.power(1.0 - cells[d.id], j)))) for d in defs)
        value = sum(content[d.id] * float(np.sum(x[d.id] * (1.0 - np.power(1.0 - cells[d.id], j)))) for d in defs)
        out.append({"cell": j, "recovery": value / total if total > 0.0 else 0.0, "grade": value / mass if mass > 0.0 else 0.0,
                    "mass_pull": mass / feed_mass if feed_mass > 0.0 else 0.0})
    return out
